Build and release wheel packages

gtristan opened the pull request:
This updates the CI config to:

  * build x86_64 wheel packages on each update of 'master', which can be downloaded as Action artifacts. The packages contain release binaries of BuildBox casd/fuse/run-bubblewrap.
  * build sdist and wheel packages on each release tag, test the wheels, and upload them to Test PyPI at https://test.pypi.org/project/BuildStream/

Before actually landing this, we need to make some changes:

  * [x] update documentation around release process and installation (https://gitlab.com/BuildStream/website/-/merge_requests/153)
  * [x] release to the real PyPI instead of [Test PyPI](https://test.pypi.org/) (in release.yml)
  * [x] create and add a [PyPI API token](https://pypi.org/help/#apitoken) as a [repository secret](https://docs.github.com/en/codespaces/managing-codespaces-for-your-organization/managing-encrypted-secrets-for-your-repository-and-organization-for-github-codespaces)
  * [x] review and merge latest [buildbox-fuse patches](https://gitlab.com/BuildGrid/buildbox/buildbox-fuse/-/merge_requests/46) so we can have a new buildbox release
  * [x] tag [buildbox-integration](https://gitlab.com/BuildGrid/buildbox/buildbox-integration/) so that we can have the binaries uploaded with the [new descriptive filenames](https://gitlab.com/BuildGrid/buildbox/buildbox-integration/-/merge_requests/5)
  * [x] apply new patch 85a8953940ee8a03d85b0277457583b67acd1656 to update the permalinks for downloading buildbox binaries

Fixes #1712

Commit:
Add option to bundle BuildBox binaries inside Python package

BuildBox is not widely distributed in binary form yet. For convience,
add a mechanism to bundle prebuilt binaries in the Python wheel
packages.

Setting BST_BUNDLE_BUILDBOX=1 when setup.py runs, causes the bundled
binaries to be included in the binary package. Its up to the packager
to make appropriate binaries available in
`src/buildstream/subprojects/buildbox/`.

BuildStream will search the package subprojects/ dir when looking for
BuildBox binaries on the host in all cases, prioritizing any bundled
binaries above other ones on the host.

Related to #1712

Commit:
Mark unstable BuildStream releases with PEP440-compatible postfix

PyPI and Pip use the PEP440 version standard, which uses explicit
'.dev', '.a', '.b' and '.rc' suffixes to mark "unstable" releases.

BuildStream uses the even/odd convention. When running `pip install`
we want Pip to prefer a stable 2.0.0 release over an unstable 2.1.0
release. To make this work, all unstable releases now have a .dev0
suffix automatically added in setup.py, e.g. `2.1.0.dev0`.

This is necessary for us to fully automate releases to PyPI from GitHub
Actions.

Commit:
docs: Update release process

Commit:
tox: Remove obsolete '-e release' mode

Commit:
.github/workflows/merge.yml: Rename 'build' and 'publish'

We are going to build and publish more than just docs, so add _docs
to these job names to make it clear.

Commit:
Build and release wheels packages as part of GitHub Actions

This updates the CI config to:

  * build wheel packages in pre-merge CI
    * download BuildBox release binaries to build the wheels
    * these can be downloaded as action artifacts
  * test built wheel packages in pre-merge CI
  * build sdist and wheel packages on each release tag, and upload
    them to Test PyPI at https://test.pypi.org/project/BuildStream/

The new workflows are based on examples at
https://cibuildwheel.readthedocs.io/en/stable/setup/, avoiding use of
the GitHub Actions from `pypa/*` as these are not permitted to be used
in apache/buildstream project.

gtristan commented:
This branch is a cleanup of #1714 and addresses some of the last review comments (because @ssssam is on vacation for the rest of the week).

This effectively replaces #1714

gtristan committed:
.github/wheel-helpers/fetch-latest-buildbox-release.sh: Fix download link

Since renaming the binary tarballs to be more descriptive upstream,
we need to fix the download link where we download the binary tarballs.

See: https://gitlab.com/BuildGrid/buildbox/buildbox-integration/-/merge_requests/5

gtristan commented:
We recently changed the name of the binaries generated by `buildbox-integration` in https://gitlab.com/BuildGrid/buildbox/buildbox-integration/-/merge_requests/5

So, we are currently waiting for:

* A new tag on `buildbox-integration` repo, so that the new binary name gets published
  * Someone please review https://gitlab.com/BuildGrid/buildbox/buildbox-fuse/-/merge_requests/46 so that this can happen now
* Me to apply the local change I'm holding onto which changes the binary permalink we download binaries from
* A pipeline to pass here with the new change applied

And then we can finally release and tag and see automatically generated and published wheels which include buildbox.

ghost commented:
The generated wheels cannot build gnome-build-meta due to https://github.com/apache/buildstream/issues/1733. We should resolve that before final release.

Commit:
Set PATH of buildbox-casd to include bundled buildbox-fuse

Otherwise, buildbox-casd falls back to its HardLinkStager backend
which does not provide sufficient capabilities to build
gnome-build-meta.

See: https://github.com/apache/buildstream/issues/1733

ghost commented:
Here's a patch that fixes buildbox-casd operation with minimal churn: https://github.com/apache/buildstream/commit/c529c615f2dbfb216d03e66c315114e5bc77b3fa

We could alternatively simplify this branch by removing all the _get_host_tool_internal() helpers and setting `os.environ["PATH"]` on startup to include `os.path.join(_site.subprojects, "buildbox")`.

gtristan commented:
> Here's a patch that fixes buildbox-casd operation with minimal churn: [c529c61](https://github.com/apache/buildstream/commit/c529c615f2dbfb216d03e66c315114e5bc77b3fa)
> 
> We could alternatively simplify this branch by removing all the _get_host_tool_internal() helpers and setting `os.environ["PATH"]` on startup to include `os.path.join(_site.subprojects, "buildbox")`.

I've applied the patch to this branch thanks.

I think I also like the patch this way because the `PATH` setting resides in codepaths which are clearly tied to launching buildbox binaries, whereas globally setting `PATH` at startup would have to live in a disconnected/unrelated codepath.

Global environment setting was unavoidable for setting `GRPC_ENABLE_FORK_SUPPORT` in cfe3d71c71 (which is unfortunatly confusing since the generated entrypoint is nowhere close to grpc usage/initialization).
